File: render/figures/composition.py
"""Where the time and the blocks went."""

from __future__ import annotations

import pandas as pd
import plotly.graph_objects as go

from copytext import DOW, t, tpl

from .. import theme
from ..theme import CATEGORY_COLOR, MONO
from .frame import frame


def category_area(cat_daily: pd.DataFrame, title: str, h: int = 360) -> go.Figure:
    """Minutes per category and day, stacked. Fixed order, never by rank."""
    order = [c for c in CATEGORY_COLOR if c in set(cat_daily["category"])]
    wide = (cat_daily.pivot_table(index="day", columns="category",
                                  values="minutes", aggfunc="sum")
            .reindex(columns=order).fillna(0))
    roll = wide.rolling(3, min_periods=1, center=True).mean()
    fig = go.Figure()
    for cat in order:
        fig.add_trace(go.Scatter(
            x=roll.index, y=roll[cat], mode="lines", stackgroup="one",
            name=t(f"category.{cat}"),
            line=dict(width=1.2, color=theme.CARD),
            fillcolor=CATEGORY_COLOR[cat],
            hovertemplate=tpl("hover.category", category=t(f"category.{cat}")),
        ))
    # the last tick is a date, which needs more room than the template's 24px
    fig.update_layout(title=title, yaxis_title=t("axis.minutes_rolling"),
                      margin=dict(t=48, r=44, b=68, l=56))
    fig.update_xaxes(tickformat="%d %b")
    return frame(fig, h)


def top_bars(tot: pd.DataFrame, title: str, n: int = 10,
             h: int = 380) -> go.Figure:
    """Horizontal ranking by minutes, coloured by category."""
    d = tot.head(n).iloc[::-1]
    fig = go.Figure(go.Bar(
        x=d["minutes"], y=d["label"], orientation="h",
        marker=dict(color=[CATEGORY_COLOR.get(c, theme.MUTED) for c in d["category"]],
                    line=dict(color=theme.CARD, width=1.5)),
        text=[t("text.minutes", minutes=m) for m in d["minutes"]],
        textposition="outside",
        textfont=dict(family=MONO, size=11, color=theme.INK_2),
        customdata=d.assign(cat=[t(f"category.{c}") for c in d["category"]])
                    [["opens", "min_per_open", "cat"]].values,
        hovertemplate=tpl("hover.top_bars"),
    ))
    fig.update_layout(title=title, xaxis_title=t("axis.minutes_month"), bargap=.28,
                      margin=dict(t=48, r=48, b=56, l=110))
    fig.update_xaxes(range=[0, d["minutes"].max() * 1.24])
    fig.update_yaxes(tickfont=dict(family=MONO, size=11, color=theme.INK))
    return frame(fig, h, legend=False)


def hour_heat(hh: pd.DataFrame, user: str, h: int = 330) -> go.Figure:
    """Weekly clock: screen minutes by weekday and hour."""
    grid = (hh.pivot_table(index="dow", columns="hour", values="minutes",
                           aggfunc="sum")
            .reindex(index=range(7), columns=range(24)).fillna(0))
    fig = go.Figure(go.Heatmap(
        z=grid.values, x=list(range(24)), y=DOW,
        colorscale=theme.HEAT,
        xgap=2, ygap=2,
        colorbar=dict(title=dict(text=t("unit.min"),
                                 font=dict(family=MONO, size=10,
                                                      color=theme.INK_2)),
                      tickfont=dict(family=MONO, size=10, color=theme.INK_2),
                      outlinewidth=0, thickness=9, len=.8, x=1.02),
        hovertemplate=tpl("hover.heat"),
    ))
    fig.update_layout(title=t("chart.hour_heat", user=user),
                      xaxis_title=t("axis.local_time"))
    fig.update_xaxes(dtick=2, showgrid=False)
    fig.update_yaxes(showgrid=False, autorange="reversed",
                     tickfont=dict(family=MONO, size=11, color=theme.INK))
    return frame(fig, h, legend=False, hovermode="closest")


# ---------------------------------------------------------------------------
# Blocks
# ---------------------------------------------------------------------------

def blocks_daily(bf: pd.DataFrame, title: str, h: int = 340) -> go.Figure:
    """Blocked attempts per day, stacked by category."""
    order = [c for c in CATEGORY_COLOR if c in set(bf["category"])]
    wide = (bf.pivot_table(index="day", columns="category", values="target",
                           aggfunc="count").reindex(columns=order).fillna(0))
    fig = go.Figure()
    for cat in order:
        fig.add_trace(go.Bar(
            x=wide.index, y=wide[cat], name=t(f"category.{cat}"),
            marker=dict(color=CATEGORY_COLOR[cat],
                        line=dict(color=theme.CARD, width=1.2)),
            hovertemplate=tpl("hover.blocks_category",
                              category=t(f"category.{cat}")),
        ))
    fig.update_layout(title=title, yaxis_title=t("axis.blocked_attempts"),
                      barmode="stack", bargap=.2)
    fig.update_xaxes(tickformat="%d %b")
    return frame(fig, h)


def blocks_by_hour(bf: pd.DataFrame, title: str, h: int = 300) -> go.Figure:
    """What time the wall gets hit; sensitive against the rest."""
    sens = bf[bf["category"].isin(["ADULT", "GAMBLING"])]
    rest = bf[~bf["category"].isin(["ADULT", "GAMBLING"])]
    fig = go.Figure()
    for name, d, color in ((t("series.ordinary"), rest, "#3987e5"),
                           (t("series.sensitive"), sens, "#e66767")):
        counts = d.groupby("hour").size().reindex(range(24)).fillna(0)
        fig.add_trace(go.Bar(
            x=list(range(24)), y=counts.values, name=name,
            marker=dict(color=color, line=dict(color=theme.CARD, width=1.2)),
            hovertemplate=tpl("hover.blocks_hour", name=name),
        ))
    fig.update_layout(title=title, xaxis_title=t("axis.local_time"),
                      yaxis_title=t("axis.attempts_month"), barmode="stack",
                      bargap=.15,
                      margin=dict(t=48, r=24, b=96, l=56),
                      legend=dict(y=-0.3))
    fig.update_xaxes(dtick=2)
    return frame(fig, h)

File: web/charts/composition.ts
/**
 * Where the time and the blocks went.
 *
 * Every stack here uses the fixed category order from the theme, never the
 * ranking: a band that grows really is growing, and has not just overtaken
 * its neighbour and swapped places with it.
 */

import { CATEGORY_LABEL, axis as unit, hover, series, title } from "../copy/figures";
import { DOW } from "../copy/figures";
import { rollingMean, thousands } from "../format";
import { CATEGORY_COLOR, CATEGORY_ORDER, MONO, type Surface } from "../theme";
import type {
  Blocks, Category, CategoryDay, HeatCell, UsageTotal, UserId,
} from "../types/index";
import { frame } from "./frame";
import type { Figure, Trace } from "./plotly";

/** The categories present in this data, in the theme's order. */
function present(found: Iterable<Category>): Category[] {
  const seen = new Set(found);
  return CATEGORY_ORDER.filter((c) => seen.has(c));
}

/** Minutes per category and day, stacked. */
export function categoryArea(s: Surface, rows: CategoryDay[],
                             heading: string): Figure {
  const days = [...new Set(rows.map((r) => r.day))].sort();
  const order = present(rows.map((r) => r.category));
  const byKey = new Map(rows.map((r) => [`${r.day}|${r.category}`, r.minutes]));

  const data: Trace[] = order.map((category) => {
    const values = days.map((day) => byKey.get(`${day}|${category}`) ?? 0);
    return {
      type: "scatter", mode: "lines", stackgroup: "one",
      x: days, y: rollingMean(values, 3, 1),
      name: CATEGORY_LABEL[category],
      line: { width: 1.2, color: s.card },
      fillcolor: CATEGORY_COLOR[category],
      hovertemplate: hover.category(CATEGORY_LABEL[category]),
    };
  });

  return frame(s, data, {
    title: { text: heading },
    yaxis: { title: { text: unit.minutesRolling } },
    xaxis: { tickformat: "%d %b" },
    // The last tick is a date, which needs more room than the template's 24.
    margin: { t: 48, r: 44, b: 68, l: 56 },
  }, 360);
}

/** Horizontal ranking by minutes, coloured by category. */
export function topBars(s: Surface, totals: UsageTotal[], heading: string,
                        n = 10): Figure {
  // Plotly draws a horizontal bar chart bottom-up, so the biggest goes last.
  const rows = totals.slice(0, n).reverse();
  const most = Math.max(...rows.map((r) => r.minutes), 1);
  return frame(s, [{
    type: "bar", orientation: "h",
    x: rows.map((r) => r.minutes), y: rows.map((r) => r.label),
    marker: {
      color: rows.map((r) => CATEGORY_COLOR[r.category] ?? s.muted),
      line: { color: s.card, width: 1.5 },
    },
    text: rows.map((r) => `${thousands(r.minutes)} min`),
    textposition: "outside",
    textfont: { family: MONO, size: 11, color: s.ink2 },
    customdata: rows.map((r) => [r.opens, r.min_per_open,
                                 CATEGORY_LABEL[r.category]]),
    hovertemplate: hover.topBars,
  }], {
    title: { text: heading },
    xaxis: { title: { text: unit.minutesMonth }, range: [0, most * 1.24] },
    yaxis: { tickfont: { family: MONO, size: 11, color: s.ink } },
    bargap: 0.28, showlegend: false,
    margin: { t: 48, r: 48, b: 56, l: 110 },
  }, 380);
}

/** Weekly clock: screen minutes by weekday and hour. */
export function hourHeat(s: Surface, cells: HeatCell[], user: UserId): Figure {
  const grid: number[][] = Array.from({ length: 7 },
                                      () => Array.from({ length: 24 }, () => 0));
  for (const cell of cells) {
    const row = grid[cell.dow];
    if (row) row[cell.hour] = (row[cell.hour] ?? 0) + cell.minutes;
  }
  return frame(s, [{
    type: "heatmap", z: grid,
    x: Array.from({ length: 24 }, (_, h) => h), y: [...DOW],
    colorscale: s.heat, xgap: 2, ygap: 2,
    colorbar: {
      title: { text: unit.min,
               font: { family: MONO, size: 10, color: s.ink2 } },
      tickfont: { family: MONO, size: 10, color: s.ink2 },
      outlinewidth: 0, thickness: 9, len: 0.8, x: 1.02,
    },
    hovertemplate: hover.heat,
  }], {
    title: { text: title.hourHeat(user) },
    xaxis: { title: { text: unit.localTime }, dtick: 2, showgrid: false },
    yaxis: { showgrid: false, autorange: "reversed",
             tickfont: { family: MONO, size: 11, color: s.ink } },
    showlegend: false, hovermode: "closest",
  }, 330);
}

// ---------------------------------------------------------------------------
// Blocks
// ---------------------------------------------------------------------------

/** Blocked attempts per day, stacked by category. */
export function blocksDaily(s: Surface, blocks: Blocks,
                            heading: string): Figure {
  const days = [...new Set(blocks.byDay.map((r) => r.day))].sort();
  const order = present(blocks.byDay.map((r) => r.category));
  const byKey = new Map(blocks.byDay.map((r) => [`${r.day}|${r.category}`,
                                                 r.count]));
  const data: Trace[] = order.map((category) => ({
    type: "bar", x: days,
    y: days.map((day) => byKey.get(`${day}|${category}`) ?? 0),
    name: CATEGORY_LABEL[category],
    marker: { color: CATEGORY_COLOR[category],
              line: { color: s.card, width: 1.2 } },
    hovertemplate: hover.blocksCategory(CATEGORY_LABEL[category]),
  }));
  return frame(s, data, {
    title: { text: heading },
    yaxis: { title: { text: unit.blockedAttempts } },
    xaxis: { tickformat: "%d %b" },
    barmode: "stack", bargap: 0.2,
  });
}

const SENSITIVE: Category[] = ["ADULT", "GAMBLING"];

/** What time the wall gets hit; sensitive against the rest. */
export function blocksByHour(s: Surface, blocks: Blocks,
                             heading: string): Figure {
  const hours = Array.from({ length: 24 }, (_, h) => h);
  const tally = (wanted: boolean): number[] => {
    const counts = hours.map(() => 0);
    for (const row of blocks.byHour) {
      if (SENSITIVE.includes(row.category) === wanted) {
        counts[row.hour] = (counts[row.hour] ?? 0) + row.count;
      }
    }
    return counts;
  };
  const data: Trace[] = [
    [series.ordinary, tally(false), "#3987e5"],
    [series.sensitive, tally(true), "#e66767"],
  ].map(([name, counts, color]) => ({
    type: "bar", x: hours, y: counts as number[], name: name as string,
    marker: { color: color as string, line: { color: s.card, width: 1.2 } },
    hovertemplate: hover.blocksHour(name as string),
  }));
  return frame(s, data, {
    title: { text: heading },
    xaxis: { title: { text: unit.localTime }, dtick: 2 },
    yaxis: { title: { text: unit.attemptsMonth } },
    barmode: "stack", bargap: 0.15,
    margin: { t: 48, r: 24, b: 96, l: 56 },
    legend: { y: -0.3 },
  }, 300);
}
